### PR TITLE
Farm: Add optional argument of original_caller

### PR DESCRIPTION
**Pull request by CostinCarabas**
Add optional argument of original_caller in farm endpoints If original_caller exists - verify that caller is the proxy contract

### CODE REVIEW

File: dex/farm/src/base_functions.rs [hunk 1]
```diff
@@ -222,4 +222,27 @@ pub trait BaseFunctionsModule:
             "May only call this function through VM query"
         );
     }
+
+    // fn require_known_proxy_from_optional(&self, opt_orig_caller: OptionalValue<ManagedAddress>) -> ManagedAddress {
```
Comment (sasurobert) [inline dex/farm/src/base_functions.rs:226]: Why commented?

File: dex/farm/src/lib.rs [hunk 2]
```diff
@@ -82,41 +83,69 @@ pub trait Farm:
...
+            OptionalValue::Some(opt_caller) => {
+                self.require_sc_address_whitelisted(&caller);
+                opt_caller
+            },
+            OptionalValue::None => caller.clone(),
+        };
+
+        let output_farm_token_payment = self.compound_rewards(&orig_caller);
+
+        // let caller = self.blockchain().get_caller();
```
Comment (sasurobert) [inline dex/farm/src/lib.rs:127]: Delete commented code.

File: dex/farm/src/lib.rs [hunk 3]
```diff
@@ -82,41 +83,69 @@ pub trait Farm:
...
-        let claim_rewards_result = self.claim_rewards(&caller);
+        let orig_caller = match opt_orig_caller {
+            OptionalValue::Some(opt_caller) => {
+                self.require_sc_address_whitelisted(&caller);
+                opt_caller
+            },
+            OptionalValue::None => caller.clone(),
+        };
+
+        let claim_rewards_result = self.claim_rewards(&orig_caller);
```
Comment (sasurobert) [inline dex/farm/src/lib.rs:104]: This will actually call the energy contract further, no?

**Review by sasurobert — CHANGES_REQUESTED** [review]

File: dex/farm/src/lib.rs (hunk 3 above)
Comment (CostinCarabas) [reply dex/farm/src/lib.rs:104]: Nope, all the claim rewards logic is inside the farm SC.

Comment (CostinCarabas) [reply dex/farm/src/lib.rs:98]: Actually, yes, this will call `farm-boosted-yields` SC.

**Review by dorin-iancu — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Pull request merged**